File: server/mosh.mjs
// Native-ffmpeg datamosh, reusing the browser engine's pure modules so the
// structural effects are byte-for-byte the same as the client. The only extra
// is `motionTransfer`, which needs FFglitch (ffgac/ffedit) — see README.
import { spawn } from 'node:child_process'
import { readFile, writeFile, mkdtemp, rm } from 'node:fs/promises'
import { tmpdir } from 'node:os'
import { join } from 'node:path'

import { parseAvi, writeAvi, frameType } from '../src/mosh/avi.ts'
import {
  bloom,
  bloomBurst,
  stutter,
  shuffle,
  sortByMotion,
  reverse,
  transition,
  applyWindowed,
} from '../src/mosh/ops.ts'

const GOP = { bloom: 12, bloomBurst: 12, stutter: 300, shuffle: 9999, sort: 9999, reverse: 9999, transition: 15 }

function run(cmd, args) {
  return new Promise((resolve, reject) => {
    const p = spawn(cmd, args, { stdio: ['ignore', 'ignore', 'pipe'] })
    let err = ''
    p.stderr.on('data', (d) => (err += d))
    p.on('error', reject)
    p.on('close', (code) => (code === 0 ? resolve() : reject(new Error(`${cmd} failed: ${err.slice(-500)}`))))
  })
}

const ffmpeg = (args) => run('ffmpeg', ['-hide_banner', '-loglevel', 'error', '-y', ...args])

async function transcodeToChunks(input, out, vf, gop) {
  await ffmpeg([
    '-i', input, '-an', '-vf', vf,
    '-c:v', 'mpeg4', '-vtag', 'xvid', '-q:v', '1',
    '-g', String(gop), '-bf', '0', '-mbd', 'rd', '-flags', '+mv4', '-trellis', '2',
    out,
  ])
  return parseAvi(await readFile(out))
}

function applySingle(effect, chunks, intensity, seed) {
  switch (effect) {
    case 'bloom': return bloom(chunks, intensity, seed)
    case 'bloomBurst': return bloomBurst(chunks, intensity, seed)
    case 'stutter': return stutter(chunks, intensity)
    case 'shuffle': return shuffle(chunks, intensity, seed)
    case 'sort': return sortByMotion(chunks)
    case 'reverse': return reverse(chunks)
    default: return chunks
  }
}

const fit = (max) => `scale='min(${max},iw)':-2:flags=bicubic`
const norm = (w, h) =>
  `scale=${w}:${h}:force_original_aspect_ratio=decrease,pad=${w}:${h}:(ow-iw)/2:(oh-ih)/2:color=black,setsar=1`

async function encode(aviPath, outPath, audioSrc, keepAudio) {
  if (keepAudio && audioSrc) {
    const video = `${outPath}.v.mp4`
    await ffmpeg(['-i', aviPath, '-c:v', 'libx264', '-crf', '18', '-pix_fmt', 'yuv420p', '-movflags', '+faststart', '-an', video])
    try {
      await ffmpeg(['-i', video, '-i', audioSrc, '-map', '0:v:0', '-map', '1:a:0?', '-c:v', 'copy', '-c:a', 'aac', '-shortest', outPath])
      return
    } catch {
      /* no audio / aac — fall through to video-only */
      await ffmpeg(['-i', video, '-c', 'copy', outPath])
      return
    }
  }
  await ffmpeg(['-i', aviPath, '-c:v', 'libx264', '-crf', '18', '-pix_fmt', 'yuv420p', '-movflags', '+faststart', '-an', outPath])
}

/**
 * FFglitch motion transfer: re-encode both clips with ffgac, then ffedit clip A
 * substituting clip B's motion vectors. Requires ffgac + ffedit on PATH.
 */
async function motionTransfer(dir, clipA, clipB, outPath) {
  const aGac = join(dir, 'a.mkv')
  const bGac = join(dir, 'b.mkv')
  await run('ffgac', ['-i', clipA, '-an', '-mpv_flags', '+nopimb+forcemv', '-qscale:v', '0', '-g', '999999', '-vcodec', 'mpeg4', '-f', 'avi', aGac])
  await run('ffgac', ['-i', clipB, '-an', '-mpv_flags', '+nopimb+forcemv', '-qscale:v', '0', '-g', '999999', '-vcodec', 'mpeg4', '-f', 'avi', bGac])
  // Export B's vectors, then apply onto A. Script provided in /server/scripts.
  const bJson = join(dir, 'b.mv.json')
  await run('ffedit', ['-i', bGac, '-f', 'mv', '-e', bJson])
  const merged = join(dir, 'merged.avi')
  await run('ffedit', ['-i', aGac, '-f', 'mv', '-a', bJson, '-o', merged])
  await encode(merged, outPath, null, false)
}

/**
 * Process a job. `files` = { clip, clip2? } absolute paths. `onProgress(r, phase)`.
 * Returns the path to the result mp4.
 */
export async function processJob(files, options, onProgress = () => {}) {
  const dir = await mkdtemp(join(tmpdir(), 'dmosh-'))
  const out = join(dir, 'out.mp4')
  try {
    const intensity = options.intensity ?? 0.8
    const maxDim = options.maxDimension ?? 1280
    const seed = options.seed ?? 0x6d6f7368
    const keepAudio = options.keepAudio ?? false
    const range = options.range

    if (options.effect === 'motionTransfer') {
      if (!files.clip2) throw new Error('motionTransfer needs two clips')
      onProgress(0.1, 'Motion transfer')
      await motionTransfer(dir, files.clip, files.clip2, out)
      onProgress(1, 'Done')
      return out
    }

    if (options.effect === 'transition') {
      if (!files.clip2) throw new Error('transition needs two clips')
      const w = Math.round(maxDim / 2) * 2
      const h = Math.round((maxDim * 9) / 16 / 2) * 2
      onProgress(0.05, 'Transcoding A')
      const a = await transcodeToChunks(files.clip, join(dir, 'a.avi'), norm(w, h), GOP.transition)
      onProgress(0.4, 'Transcoding B')
      const b = await transcodeToChunks(files.clip2, join(dir, 'b.avi'), norm(w, h), GOP.transition)
      onProgress(0.8, 'Encoding')
      const moshed = writeAvi(a.head, transition(a.chunks, b.chunks))
      const m = join(dir, 'm.avi')
      await writeFile(m, Buffer.from(moshed))
      await encode(m, out, files.clip, keepAudio)
      onProgress(1, 'Done')
      return out
    }

    const partial = range && !(range.start <= 0 && range.end >= 1)
    const gop = partial ? 12 : GOP[options.effect]
    onProgress(0.05, 'Transcoding')
    const { head, chunks } = await transcodeToChunks(files.clip, join(dir, 'work.avi'), fit(maxDim), gop)
    onProgress(0.6, 'Encoding')
    const moshed = writeAvi(head, applyWindowed((cs) => applySingle(options.effect, cs, intensity, seed), chunks, range))
    const m = join(dir, 'm.avi')
    await writeFile(m, Buffer.from(moshed))
    await encode(m, out, files.clip, keepAudio)
    onProgress(1, 'Done')
    return out
  } catch (e) {
    await rm(dir, { recursive: true, force: true }).catch(() => {})
    throw e
  }
}

// re-export so the server can validate effect names
export const SINGLE_EFFECTS = ['bloom', 'bloomBurst', 'stutter', 'shuffle', 'sort', 'reverse']
export { frameType }

File: crates/dmosh-core/src/mosh.rs
//! Native datamosh orchestration: ffmpeg transcode → frame surgery → re-encode.
//! Mirrors `src/mosh/datamosh.ts` but shells out to a native ffmpeg (no wasm
//! ceiling, multithreaded, libx264 — we distribute directly, so GPL is fine).

use std::path::{Path, PathBuf};
use std::process::Command;

use serde::Deserialize;

use crate::avi::{parse_avi, write_avi, Chunk};
use crate::ops::{
    apply_windowed, bloom, bloom_burst, reverse, shuffle, sort_by_motion, stutter, transition,
    DropTargets,
};

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MoshOptions {
    pub effect: String,
    #[serde(default = "default_intensity")]
    pub intensity: f32,
    #[serde(default = "default_max_dim")]
    pub max_dimension: u32,
    #[serde(default)]
    pub keep_audio: bool,
    #[serde(default = "default_seed")]
    pub seed: u32,
    #[serde(default)]
    pub range: Option<(f32, f32)>,
    #[serde(default = "default_true")]
    pub drop_i: bool,
    #[serde(default)]
    pub drop_p: bool,
    #[serde(default)]
    pub repeat: Option<u32>,
}

fn default_intensity() -> f32 {
    0.7
}
fn default_max_dim() -> u32 {
    1280
}
fn default_seed() -> u32 {
    0x6d6f_7368
}
fn default_true() -> bool {
    true
}

fn clamp01(x: f32) -> f32 {
    x.max(0.0).min(1.0)
}
fn gop_for(effect: &str) -> u32 {
    if effect == "transition" {
        30
    } else {
        9999
    }
}
fn qscale_for(intensity: f32) -> u32 {
    4 + (clamp01(intensity) * 5.0).round() as u32
}

/// Progress sink: `(0..1, phase)`.
pub type Progress<'a> = dyn Fn(f32, &str) + 'a;

fn ffmpeg_bin() -> String {
    std::env::var("DMOSH_FFMPEG").unwrap_or_else(|_| "ffmpeg".to_string())
}

fn ff(args: &[&str]) -> Result<(), String> {
    let bin = ffmpeg_bin();
    let output = Command::new(&bin)
        .args(["-hide_banner", "-loglevel", "error", "-y"])
        .args(args)
        .output()
        .map_err(|e| format!("could not run ffmpeg ('{bin}'): {e}"))?;
    if !output.status.success() {
        let tail = String::from_utf8_lossy(&output.stderr);
        let tail = &tail[tail.len().saturating_sub(600)..];
        return Err(format!("ffmpeg failed: {tail}"));
    }
    Ok(())
}

fn fit_filter(max: u32) -> String {
    format!("scale='min({max},iw)':-2:flags=bicubic")
}
fn normalize_filter(w: u32, h: u32) -> String {
    format!(
        "scale={w}:{h}:force_original_aspect_ratio=decrease,pad={w}:{h}:(ow-iw)/2:(oh-ih)/2:color=black,setsar=1"
    )
}

fn transcode_to_chunks(
    input: &Path,
    out_avi: &Path,
    vfilter: &str,
    gop: u32,
    qscale: u32,
) -> Result<crate::avi::ParsedAvi, String> {
    ff(&[
        "-i",
        &input.to_string_lossy(),
        "-an",
        "-vf",
        vfilter,
        "-c:v",
        "mpeg4",
        "-vtag",
        "xvid",
        "-qscale:v",
        &qscale.to_string(),
        "-g",
        &gop.to_string(),
        "-bf",
        "0",
        &out_avi.to_string_lossy(),
    ])?;
    let bytes = std::fs::read(out_avi).map_err(|e| e.to_string())?;
    parse_avi(&bytes)
}

fn encode_to_mp4(
    avi_path: &Path,
    out_path: &Path,
    keep_audio: bool,
    audio_src: &Path,
) -> Result<(), String> {
    if keep_audio {
        let video = avi_path.with_extension("v.mp4");
        ff(&[
            "-i", &avi_path.to_string_lossy(),
            "-c:v", "libx264", "-crf", "18", "-preset", "medium",
            "-pix_fmt", "yuv420p", "-movflags", "+faststart", "-an",
            &video.to_string_lossy(),
        ])?;
        // `1:a:0?` keeps the audio map optional — silent clips pass through.
        let muxed = ff(&[
            "-i", &video.to_string_lossy(),
            "-i", &audio_src.to_string_lossy(),
            "-map", "0:v:0", "-map", "1:a:0?",
            "-c:v", "copy", "-c:a", "aac", "-shortest",
            &out_path.to_string_lossy(),
        ]);
        if muxed.is_err() {
            std::fs::copy(&video, out_path).map_err(|e| e.to_string())?;
        }
        return Ok(());
    }
    ff(&[
        "-i", &avi_path.to_string_lossy(),
        "-c:v", "libx264", "-crf", "18", "-preset", "medium",
        "-pix_fmt", "yuv420p", "-movflags", "+faststart", "-an",
        &out_path.to_string_lossy(),
    ])
}

fn apply_single(effect: &str, chunks: &[Chunk], opts: &MoshOptions) -> Vec<Chunk> {
    match effect {
        "bloom" => bloom(chunks, opts.intensity),
        "bloomBurst" => bloom_burst(chunks, opts.intensity),
        "stutter" => stutter(chunks, opts.intensity, opts.repeat),
        "shuffle" => shuffle(chunks, opts.intensity, opts.seed),
        "sort" => sort_by_motion(chunks),
        "reverse" => reverse(chunks),
        _ => chunks.to_vec(),
    }
}

fn unique_workdir() -> Result<PathBuf, String> {
    let nanos = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map(|d| d.as_nanos())
        .unwrap_or(0);
    let dir = std::env::temp_dir().join(format!("dmosh-{}-{}", std::process::id(), nanos));
    std::fs::create_dir_all(&dir).map_err(|e| e.to_string())?;
    Ok(dir)
}

/// Datamosh one or two clips with native ffmpeg. Returns the output MP4 path.
pub fn mosh(inputs: &[PathBuf], opts: &MoshOptions, progress: &Progress) -> Result<PathBuf, String> {
    if opts.effect == "flow" {
        return Err("the Flow effect runs in the app's GPU preview, not the native engine".into());
    }
    if inputs.is_empty() {
        return Err("no input clip".into());
    }

    let dir = unique_workdir()?;
    let out = dir.join("out.mp4");
    let gop = gop_for(&opts.effect);
    let qscale = qscale_for(opts.intensity);
    let targets = DropTargets { i: opts.drop_i, p: opts.drop_p };

    if opts.effect == "transition" {
        if inputs.len() < 2 {
            return Err("transition needs two clips".into());
        }
        let w = (opts.max_dimension / 2) * 2;
        let h = (((opts.max_dimension as f32 * 9.0 / 16.0) / 2.0).round() as u32) * 2;
        let vf = normalize_filter(w, h);
        progress(0.05, "transcoding clip A");
        let a = transcode_to_chunks(&inputs[0], &dir.join("a.avi"), &vf, gop, qscale)?;
        progress(0.4, "transcoding clip B");
        let b = transcode_to_chunks(&inputs[1], &dir.join("b.avi"), &vf, gop, qscale)?;
        progress(0.8, "moshing");
        let moshed = write_avi(&a.head, &transition(&a.chunks, &b.chunks, targets, opts.intensity));
        let m = dir.join("m.avi");
        std::fs::write(&m, &moshed).map_err(|e| e.to_string())?;
        progress(0.85, "encoding");
        encode_to_mp4(&m, &out, opts.keep_audio, &inputs[0])?;
        progress(1.0, "done");
        return Ok(out);
    }

    progress(0.05, "transcoding");
    let parsed = transcode_to_chunks(&inputs[0], &dir.join("work.avi"), &fit_filter(opts.max_dimension), gop, qscale)?;
    progress(0.55, "moshing");
    let moshed_chunks = apply_windowed(
        |cs| apply_single(&opts.effect, cs, opts),
        &parsed.chunks,
        opts.range,
    );
    let moshed = write_avi(&parsed.head, &moshed_chunks);
    let m = dir.join("m.avi");
    std::fs::write(&m, &moshed).map_err(|e| e.to_string())?;
    progress(0.6, "encoding");
    encode_to_mp4(&m, &out, opts.keep_audio, &inputs[0])?;
    progress(1.0, "done");
    Ok(out)
}
